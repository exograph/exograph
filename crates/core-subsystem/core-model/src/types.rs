// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

//! Common support for representing GraphQL types such as `Int`, `List<Int>`, `Optional<Int>`, `Optional<List<Int>>`, etc.
//!

use async_graphql_parser::types::{BaseType, Type};
use async_graphql_value::Name;
use serde::{Deserialize, Serialize};

use crate::mapped_arena::{SerializableSlab, SerializableSlabIndex};

/// A type that can be used as a type for fields and return types
/// Currently supports only list and optional decorations
#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum FieldType<T> {
    Plain(T),
    List(Box<FieldType<T>>),
    Optional(Box<FieldType<T>>),
}

pub trait Named {
    fn name(&self) -> &str;
}

impl<T> FieldType<T> {
    /// Return the inner type, if any
    pub fn inner(&self) -> Option<&FieldType<T>> {
        match self {
            FieldType::Plain(_) => None,
            FieldType::List(inner) | FieldType::Optional(inner) => Some(inner),
        }
    }

    /// Return the base type (i.e. by removing optional decoration)
    pub fn base_type(&self) -> &FieldType<T> {
        match self {
            FieldType::Optional(inner_typ) => inner_typ.as_ref(),
            _ => self,
        }
    }

    /// Return the innermost (i.e. leaf) type
    pub fn innermost(&self) -> &T {
        match self {
            FieldType::Plain(inner) => inner,
            FieldType::List(inner) | FieldType::Optional(inner) => inner.innermost(),
        }
    }

    /// Wrap the destination type to match the structure of the source type
    pub fn wrap<D>(&self, dest: D) -> FieldType<D> {
        match self {
            FieldType::Plain(_) => FieldType::Plain(dest),
            FieldType::Optional(optional) => FieldType::Optional(Box::new(optional.wrap(dest))),
            FieldType::List(list) => FieldType::List(Box::new(list.wrap(dest))),
        }
    }
}

impl<T: Clone> FieldType<T> {
    /// Compute the optional version of the given type
    pub fn optional(&self) -> Self {
        match self {
            FieldType::Optional(_) => self.clone(), // Already optional
            _ => FieldType::Optional(Box::new(self.clone())),
        }
    }
}

/// Transforms the type into an introspection type
///
/// The complexity of this function is due to the fact that the GraphQL spec and hence the
/// introspection type ([`Type`]) does not support nested optionals. However, [`FieldType`]
/// being more general, does. This function will panic if it encounters a nested optional.
impl<T: Named> From<&FieldType<T>> for Type {
    fn from(ft: &FieldType<T>) -> Self {
        match ft {
            FieldType::Plain(base) => Type {
                base: BaseType::Named(Name::new(base.name())),
                nullable: false,
            },
            FieldType::Optional(underlying) => {
                let Type { base, nullable } = underlying.as_ref().into();

                if nullable {
                    panic!("Optional type cannot be nested")
                }
                Type {
                    base,
                    nullable: true,
                }
            }
            FieldType::List(underlying) => Type {
                base: BaseType::List(Box::new(underlying.as_ref().into())),
                nullable: false,
            },
        }
    }
}

impl<T: Named> Named for FieldType<T> {
    fn name(&self) -> &str {
        match self {
            FieldType::Plain(plain) => plain.name(),
            FieldType::List(list) => list.name(),
            FieldType::Optional(optional) => optional.name(),
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct BaseOperationReturnType<T> {
    pub associated_type_id: SerializableSlabIndex<T>,
    pub type_name: String,
}

impl<T> Named for BaseOperationReturnType<T> {
    fn name(&self) -> &str {
        &self.type_name
    }
}

pub type OperationReturnType<T> = FieldType<BaseOperationReturnType<T>>;

impl<T> OperationReturnType<T> {
    pub fn typ<'a>(&'a self, types: &'a SerializableSlab<T>) -> &T {
        match self {
            OperationReturnType::Plain(BaseOperationReturnType {
                associated_type_id, ..
            }) => &types[*associated_type_id],
            OperationReturnType::List(underlying) | OperationReturnType::Optional(underlying) => {
                underlying.typ(types)
            }
        }
    }

    pub fn type_name(&self) -> &str {
        match self {
            OperationReturnType::Plain(BaseOperationReturnType { type_name, .. }) => type_name,
            OperationReturnType::List(underlying) | OperationReturnType::Optional(underlying) => {
                underlying.type_name()
            }
        }
    }
}
