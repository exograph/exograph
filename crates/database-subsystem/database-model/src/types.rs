use super::access::Access;
use super::{column_id::ColumnId, relation::DatabaseRelation};
use crate::model::ModelDatabaseSystem;
use crate::operation::{DatabaseQuery, DatabaseQueryParameter};
use async_graphql_parser::types::{
    BaseType, FieldDefinition, InputObjectType, InputValueDefinition, ObjectType, Type,
    TypeDefinition, TypeKind,
};
use async_graphql_value::Name;
use core_model::mapped_arena::{SerializableSlab, SerializableSlabIndex};

use core_model::type_normalization::{
    default_positioned, default_positioned_name, FieldDefinitionProvider, InputValueProvider,
    TypeDefinitionProvider, TypeModifier,
};
use payas_sql::PhysicalTable;

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DatabaseType {
    pub name: String,
    pub plural_name: String,
    pub kind: DatabaseTypeKind,
    pub is_input: bool, // Is this to be used as an input field (such as an argument in a mutation)? Needed for introspection
}

impl DatabaseType {
    pub fn model_fields(&self) -> Vec<&DatabaseField> {
        match &self.kind {
            DatabaseTypeKind::Primitive => vec![],
            DatabaseTypeKind::Composite(DatabaseCompositeType { fields, .. }) => {
                fields.iter().collect()
            }
        }
    }

    pub fn model_field(&self, name: &str) -> Option<&DatabaseField> {
        self.model_fields()
            .into_iter()
            .find(|model_field| model_field.name == name)
    }

    pub fn pk_field(&self) -> Option<&DatabaseField> {
        match &self.kind {
            DatabaseTypeKind::Primitive => None,
            DatabaseTypeKind::Composite(composite_type) => composite_type.pk_field(),
        }
    }

    pub fn pk_column_id(&self) -> Option<ColumnId> {
        match &self.kind {
            DatabaseTypeKind::Primitive => None,
            DatabaseTypeKind::Composite(composite_type) => composite_type.pk_column_id(),
        }
    }

    pub fn table_id(&self) -> Option<SerializableSlabIndex<PhysicalTable>> {
        match &self.kind {
            DatabaseTypeKind::Composite(DatabaseCompositeType { table_id, .. }) => Some(*table_id),
            _ => None,
        }
    }

    pub fn is_primitive(&self) -> bool {
        matches!(&self.kind, DatabaseTypeKind::Primitive)
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[allow(clippy::large_enum_variant)]
pub enum DatabaseTypeKind {
    Primitive,
    Composite(DatabaseCompositeType),
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DatabaseCompositeType {
    pub fields: Vec<DatabaseField>,
    pub table_id: SerializableSlabIndex<PhysicalTable>,
    pub pk_query: SerializableSlabIndex<DatabaseQuery>,
    pub collection_query: SerializableSlabIndex<DatabaseQuery>,
    pub access: Access,
}

impl DatabaseCompositeType {
    pub fn get_field_by_name(&self, name: &str) -> Option<&DatabaseField> {
        self.fields.iter().find(|field| field.name == name)
    }

    pub fn pk_field(&self) -> Option<&DatabaseField> {
        self.fields
            .iter()
            .find(|field| matches!(&field.relation, DatabaseRelation::Pk { .. }))
    }

    pub fn pk_column_id(&self) -> Option<ColumnId> {
        self.pk_field()
            .and_then(|pk_field| pk_field.relation.self_column())
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
pub enum DatabaseTypeModifier {
    Optional,
    NonNull,
    List,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DatabaseField {
    pub name: String,
    pub typ: DatabaseFieldType,
    pub relation: DatabaseRelation,
    pub has_default_value: bool, // does this field have a default value?
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum DatabaseFieldType {
    Optional(Box<DatabaseFieldType>),
    Reference {
        type_id: SerializableSlabIndex<DatabaseType>,
        is_primitive: bool, // A way to know which arena to look up the type in
        type_name: String,
    },
    List(Box<DatabaseFieldType>),
}

impl DatabaseFieldType {
    pub fn type_id(&self) -> &SerializableSlabIndex<DatabaseType> {
        match self {
            DatabaseFieldType::Optional(underlying) | DatabaseFieldType::List(underlying) => {
                underlying.type_id()
            }
            DatabaseFieldType::Reference { type_id, .. } => type_id,
        }
    }

    pub fn is_primitive(&self) -> bool {
        match self {
            DatabaseFieldType::Optional(underlying) | DatabaseFieldType::List(underlying) => {
                underlying.is_primitive()
            }
            DatabaseFieldType::Reference { is_primitive, .. } => *is_primitive,
        }
    }

    pub fn base_type<'a>(&self, types: &'a SerializableSlab<DatabaseType>) -> &'a DatabaseType {
        match self {
            DatabaseFieldType::Optional(underlying) | DatabaseFieldType::List(underlying) => {
                underlying.base_type(types)
            }
            DatabaseFieldType::Reference { type_id, .. } => &types[*type_id],
        }
    }

    pub fn type_name(&self) -> &str {
        match self {
            DatabaseFieldType::Optional(underlying) | DatabaseFieldType::List(underlying) => {
                underlying.type_name()
            }
            DatabaseFieldType::Reference { type_name, .. } => type_name,
        }
    }

    pub fn optional(&self) -> Self {
        match self {
            DatabaseFieldType::Optional(_) => self.clone(),
            _ => DatabaseFieldType::Optional(Box::new(self.clone())),
        }
    }
}

impl TypeDefinitionProvider<ModelDatabaseSystem> for DatabaseType {
    fn type_definition(&self, system: &ModelDatabaseSystem) -> TypeDefinition {
        match &self.kind {
            DatabaseTypeKind::Primitive => TypeDefinition {
                extend: false,
                description: None,
                name: default_positioned_name(&self.name),
                directives: vec![],
                kind: TypeKind::Scalar,
            },
            DatabaseTypeKind::Composite(DatabaseCompositeType {
                fields: model_fields,
                ..
            }) => {
                let kind = if self.is_input {
                    let fields = model_fields
                        .iter()
                        .map(|model_field| default_positioned(model_field.input_value()))
                        .collect();
                    TypeKind::InputObject(InputObjectType { fields })
                } else {
                    let fields: Vec<_> = model_fields
                        .iter()
                        .map(|model_field| default_positioned(model_field.field_definition(system)))
                        .collect();

                    TypeKind::Object(ObjectType {
                        implements: vec![],
                        fields,
                    })
                };
                TypeDefinition {
                    extend: false,
                    description: None,
                    name: default_positioned_name(&self.name),
                    directives: vec![],
                    kind,
                }
            }
        }
    }
}

impl FieldDefinitionProvider<ModelDatabaseSystem> for DatabaseField {
    fn field_definition(&self, system: &ModelDatabaseSystem) -> FieldDefinition {
        let field_type = default_positioned(compute_type(&self.typ));

        let arguments = match self.relation {
            DatabaseRelation::Pk { .. }
            | DatabaseRelation::Scalar { .. }
            | DatabaseRelation::ManyToOne { .. } => {
                vec![]
            }
            DatabaseRelation::OneToMany { other_type_id, .. } => {
                let other_type = &system.database_types[other_type_id];
                match &other_type.kind {
                    DatabaseTypeKind::Primitive => panic!(),
                    DatabaseTypeKind::Composite(kind) => {
                        let collection_query = kind.collection_query;
                        let collection_query = &system.queries[collection_query];

                        let DatabaseQueryParameter {
                            predicate_param,
                            order_by_param,
                            limit_param,
                            offset_param,
                        } = &collection_query.parameter;

                        let predicate_parameter_arg =
                            predicate_param.as_ref().map(|p| p.input_value());
                        let order_by_parameter_arg =
                            order_by_param.as_ref().map(|p| p.input_value());
                        let limit_arg = limit_param.as_ref().map(|p| p.input_value());
                        let offset_arg = offset_param.as_ref().map(|p| p.input_value());

                        vec![
                            predicate_parameter_arg,
                            order_by_parameter_arg,
                            limit_arg,
                            offset_arg,
                        ]
                        .into_iter()
                        .flatten()
                        .map(default_positioned)
                        .collect()
                    }
                }
            }
        };

        FieldDefinition {
            description: None,
            name: default_positioned_name(&self.name),
            arguments,
            ty: field_type,
            directives: vec![],
        }
    }
}

pub fn compute_type(typ: &DatabaseFieldType) -> Type {
    fn compute_base_type(typ: &DatabaseFieldType) -> BaseType {
        match typ {
            DatabaseFieldType::Optional(underlying) => compute_base_type(underlying),
            DatabaseFieldType::Reference { type_name, .. } => BaseType::Named(Name::new(type_name)),
            DatabaseFieldType::List(underlying) => {
                BaseType::List(Box::new(compute_type(underlying)))
            }
        }
    }

    match typ {
        DatabaseFieldType::Optional(underlying) => Type {
            base: compute_base_type(underlying),
            nullable: true,
        },
        DatabaseFieldType::Reference { type_name, .. } => Type {
            base: BaseType::Named(Name::new(type_name)),
            nullable: false,
        },
        DatabaseFieldType::List(underlying) => Type {
            base: BaseType::List(Box::new(compute_type(underlying))),
            nullable: false,
        },
    }
}

impl From<&DatabaseTypeModifier> for TypeModifier {
    fn from(modifier: &DatabaseTypeModifier) -> Self {
        match modifier {
            DatabaseTypeModifier::Optional => TypeModifier::Optional,
            DatabaseTypeModifier::NonNull => TypeModifier::NonNull,
            DatabaseTypeModifier::List => TypeModifier::List,
        }
    }
}

// We need to a special case for the GqlField type, so that we can properly
// created nested types such as Optional(List(List(String))). The blanket impl
// above will not work for nested types like these.
impl InputValueProvider for DatabaseField {
    fn input_value(&self) -> InputValueDefinition {
        let field_type = default_positioned(compute_type(&self.typ));

        InputValueDefinition {
            description: None,
            name: default_positioned_name(&self.name),
            ty: field_type,
            default_value: None,
            directives: vec![],
        }
    }
}
