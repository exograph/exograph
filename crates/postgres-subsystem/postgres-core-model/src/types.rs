// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use super::relation::PostgresRelation;
use crate::aggregate::AggregateField;
use crate::vector_distance::VectorDistanceField;

use common::value::Val;
use core_plugin_interface::core_model::context_type::ContextSelection;
use core_plugin_interface::core_model::types::TypeValidation;
use core_plugin_interface::core_model::{
    mapped_arena::{SerializableSlab, SerializableSlabIndex},
    types::{FieldType, Named},
};

use crate::access::Access;

use exo_sql::PhysicalTable;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone, Copy)]
pub enum TypeIndex<CT> {
    Primitive(SerializableSlabIndex<PostgresPrimitiveType>),
    Composite(SerializableSlabIndex<CT>),
}

impl<CT> TypeIndex<CT> {
    pub fn to_type<'a>(
        &self,
        primitive_types: &'a SerializableSlab<PostgresPrimitiveType>,
        entity_types: &'a SerializableSlab<CT>,
    ) -> PostgresType<'a, CT> {
        match self {
            TypeIndex::Primitive(index) => PostgresType::Primitive(&primitive_types[*index]),
            TypeIndex::Composite(index) => PostgresType::Composite(&entity_types[*index]),
        }
    }
}

#[derive(Debug)]
pub enum PostgresType<'a, CT> {
    Primitive(&'a PostgresPrimitiveType),
    Composite(&'a CT),
}

impl<'a, CT: Named> PostgresType<'a, CT> {
    pub fn name(&self) -> &str {
        match self {
            PostgresType::Primitive(primitive_type) => &primitive_type.name,
            PostgresType::Composite(composite_type) => composite_type.name(),
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PostgresPrimitiveType {
    pub name: String,
}

#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq)]
pub enum EntityRepresentation {
    Json,
    Managed,
    NotManaged,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct EntityType {
    pub name: String,
    pub plural_name: String,
    pub representation: EntityRepresentation,

    pub fields: Vec<PostgresField<EntityType>>,
    pub agg_fields: Vec<AggregateField>,
    pub vector_distance_fields: Vec<VectorDistanceField>,

    pub table_id: SerializableSlabIndex<PhysicalTable>,
    pub access: Access,
}

/// Encapsulates a field on an entity type (mirros how `ColumnId` is structured)
#[derive(Serialize, Deserialize, Clone, Copy, Debug)]
pub struct EntityFieldId(usize, SerializableSlabIndex<EntityType>);

impl EntityFieldId {
    pub fn entity_type_id(&self) -> SerializableSlabIndex<EntityType> {
        self.1
    }

    pub fn resolve<'a>(
        &self,
        types: &'a SerializableSlab<EntityType>,
    ) -> &'a PostgresField<EntityType> {
        let entity = &types[self.1];
        &entity.fields[self.0]
    }
}

impl Named for EntityType {
    fn name(&self) -> &str {
        &self.name
    }
}

impl EntityType {
    pub fn field_by_name(&self, name: &str) -> Option<&PostgresField<EntityType>> {
        self.fields.iter().find(|field| field.name == name)
    }

    pub fn pk_fields(&self) -> Vec<&PostgresField<EntityType>> {
        self.fields
            .iter()
            .filter(|field| field.relation.is_pk())
            .collect()
    }

    pub fn pk_field_ids(&self, entity_id: SerializableSlabIndex<EntityType>) -> Vec<EntityFieldId> {
        self.fields
            .iter()
            .enumerate()
            .filter(|(_, field)| field.relation.is_pk())
            .map(|(field_index, _)| EntityFieldId(field_index, entity_id))
            .collect()
    }

    pub fn aggregate_field_by_name(&self, name: &str) -> Option<&AggregateField> {
        self.agg_fields.iter().find(|field| field.name == name)
    }

    pub fn vector_distance_field_by_name(&self, name: &str) -> Option<&VectorDistanceField> {
        self.vector_distance_fields
            .iter()
            .find(|field| field.name == name)
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum PostgresFieldDefaultValue {
    Static(Val),
    Dynamic(ContextSelection),
    Function(String), // Postgres function name such as `now()`
    AutoIncrement,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct PostgresField<CT> {
    pub name: String,
    pub typ: FieldType<PostgresFieldType<CT>>,
    pub relation: PostgresRelation,
    pub default_value: Option<PostgresFieldDefaultValue>,
    pub readonly: bool,
    pub access: Access,
    pub type_validation: Option<TypeValidation>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct PostgresFieldType<CT> {
    pub type_id: TypeIndex<CT>,
    pub type_name: String,
}

impl<CT> Named for PostgresFieldType<CT> {
    fn name(&self) -> &str {
        &self.type_name
    }
}

pub fn base_type<'a, CT>(
    typ: &FieldType<PostgresFieldType<CT>>,
    primitive_types: &'a SerializableSlab<PostgresPrimitiveType>,
    entity_types: &'a SerializableSlab<CT>,
) -> PostgresType<'a, CT> {
    typ.innermost()
        .type_id
        .to_type(primitive_types, entity_types)
}
