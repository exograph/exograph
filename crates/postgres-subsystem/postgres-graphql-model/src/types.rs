// Copyright Exograph, Inc. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file at the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use crate::operation::{OperationParameters, PostgresOperation};
use crate::query::CollectionQueryParameters;
use crate::subsystem::PostgresGraphQLSubsystem;
use async_graphql_parser::types::{
    FieldDefinition, InputObjectType, ObjectType, Type, TypeDefinition, TypeKind,
};
use core_plugin_interface::core_model::access::AccessPredicateExpression;
use core_plugin_interface::core_model::primitive_type::vector_introspection_base_type;
use core_plugin_interface::core_model::types::{DirectivesProvider, TypeValidation};
use core_plugin_interface::core_model::{
    mapped_arena::SerializableSlabIndex,
    type_normalization::{
        default_positioned, default_positioned_name, FieldDefinitionProvider, InputValueProvider,
        Parameter, TypeDefinitionProvider,
    },
    types::{FieldType, Named},
};
use postgres_core_model::relation::OneToManyRelation;
use postgres_core_model::relation::PostgresRelation;

use postgres_core_model::access::{
    DatabaseAccessPrimitiveExpression, InputAccessPrimitiveExpression,
};

use postgres_core_model::types::{EntityType, PostgresField, PostgresPrimitiveType};
use serde::{Deserialize, Serialize};

/// Mutation input types such as `CreatePostInput` and `UpdatePostInput`
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MutationType {
    pub name: String,
    pub fields: Vec<PostgresField<MutationType>>,
    pub entity_id: SerializableSlabIndex<EntityType>,

    pub input_access:
        Option<SerializableSlabIndex<AccessPredicateExpression<InputAccessPrimitiveExpression>>>,
    pub database_access:
        Option<SerializableSlabIndex<AccessPredicateExpression<DatabaseAccessPrimitiveExpression>>>,
}

impl TypeDefinitionProvider<PostgresGraphQLSubsystem> for PostgresPrimitiveType {
    fn type_definition(&self, _system: &PostgresGraphQLSubsystem) -> TypeDefinition {
        TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name(&self.name),
            directives: vec![],
            kind: TypeKind::Scalar,
        }
    }
}

impl TypeDefinitionProvider<PostgresGraphQLSubsystem> for EntityType {
    fn type_definition(&self, system: &PostgresGraphQLSubsystem) -> TypeDefinition {
        let EntityType {
            fields,
            agg_fields,
            vector_distance_fields,
            ..
        } = self;

        let kind = {
            let entity = fields
                .iter()
                .map(|field| default_positioned(field.field_definition(system)));

            let agg_fields = agg_fields
                .iter()
                .map(|field| default_positioned(field.field_definition(system)));

            let vector_distance_fields = vector_distance_fields
                .iter()
                .map(|field| default_positioned(field.field_definition(system)));

            let fields = entity
                .chain(agg_fields)
                .chain(vector_distance_fields)
                .collect();

            TypeKind::Object(ObjectType {
                implements: vec![],
                fields,
            })
        };
        TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name(&self.name),
            directives: vec![],
            kind,
        }
    }
}

impl TypeDefinitionProvider<PostgresGraphQLSubsystem> for MutationType {
    fn type_definition(&self, _system: &PostgresGraphQLSubsystem) -> TypeDefinition {
        let kind = {
            let fields = self
                .fields
                .iter()
                .flat_map(|field| {
                    (!field.readonly).then_some(default_positioned(
                        PostgresMutationField(field).input_value(),
                    ))
                })
                .collect();
            TypeKind::InputObject(InputObjectType { fields })
        };
        TypeDefinition {
            extend: false,
            description: None,
            name: default_positioned_name(&self.name),
            directives: vec![],
            kind,
        }
    }
}

impl<CT> FieldDefinitionProvider<PostgresGraphQLSubsystem> for PostgresField<CT> {
    fn field_definition(&self, system: &PostgresGraphQLSubsystem) -> FieldDefinition {
        let field_type = default_positioned((&self.typ).into());
        let mut directives = vec![];

        if let Some(type_validation) = &self.type_validation {
            directives = type_validation
                .get_directives()
                .iter()
                .map(|d| default_positioned(d.to_owned()))
                .collect();
        }

        // Special case for Vector. Even though it is a "scalar" from the perspective of the
        // database, it is a list of floats from the perspective of the GraphQL schema.
        // TODO: This should be handled in a more general way (probably best done with https://github.com/exograph/exograph/issues/603)
        if self.typ.base_type().name() == "Vector" {
            let base_list_type = vector_introspection_base_type();

            return FieldDefinition {
                description: None,
                name: default_positioned_name(&self.name),
                arguments: vec![],
                ty: default_positioned(Type {
                    base: base_list_type,
                    nullable: matches!(self.typ, FieldType::Optional(_)),
                }),
                directives,
            };
        }

        let arguments = match self.relation {
            PostgresRelation::Pk { .. }
            | PostgresRelation::Scalar { .. }
            | PostgresRelation::ManyToOne { .. } => {
                vec![]
            }
            PostgresRelation::OneToMany(OneToManyRelation {
                foreign_field_id, ..
            }) => {
                let foreign_type_id = foreign_field_id.entity_type_id();
                let collection_query = system.get_collection_query(foreign_type_id);

                let CollectionQueryParameters {
                    predicate_param,
                    order_by_param,
                    limit_param,
                    offset_param,
                } = &collection_query.parameters;

                [
                    predicate_param.input_value(),
                    order_by_param.input_value(),
                    limit_param.input_value(),
                    offset_param.input_value(),
                ]
                .into_iter()
                .map(default_positioned)
                .collect()
            }
        };

        FieldDefinition {
            description: None,
            name: default_positioned_name(&self.name),
            arguments,
            ty: field_type,
            directives,
        }
    }
}

// To get around the orphan rule, we wrap the field in a struct
struct PostgresMutationField<'a>(&'a PostgresField<MutationType>);

impl<'a> Parameter for PostgresMutationField<'a> {
    fn name(&self) -> &str {
        &self.0.name
    }

    fn typ(&self) -> Type {
        (&self.0.typ).into()
    }

    fn type_validation(&self) -> Option<TypeValidation> {
        self.0.type_validation.clone()
    }
}

pub trait Operation {
    fn name(&self) -> &String;
    fn parameters(&self) -> Vec<&dyn Parameter>;
    fn return_type(&self) -> Type;
}

impl<S, P: OperationParameters> FieldDefinitionProvider<S> for PostgresOperation<P> {
    fn field_definition(&self, _system: &S) -> FieldDefinition {
        let fields = self
            .parameters()
            .iter()
            .map(|parameter| default_positioned(parameter.input_value()))
            .collect();

        FieldDefinition {
            description: None,
            name: default_positioned_name(self.name()),
            arguments: fields,
            directives: vec![],
            ty: default_positioned(self.return_type()),
        }
    }
}
