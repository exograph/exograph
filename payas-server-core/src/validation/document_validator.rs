use async_graphql_parser::types::ExecutableDocument;
use serde_json::{Map, Value};

use crate::{error::ExecutionError, introspection::schema::Schema};

use super::{document::ValidatedDocument, operation_validator::OperationValidator};

/// Context for validating a document.
pub struct DocumentValidator<'a> {
    schema: &'a Schema,
    operation_name: Option<String>,
    variables: Option<Map<String, Value>>,
}

impl<'a> DocumentValidator<'a> {
    pub fn new(
        schema: &'a Schema,
        operation_name: Option<String>,
        variables: Option<Map<String, Value>>,
    ) -> Self {
        Self {
            schema,
            operation_name,
            variables,
        }
    }

    /// Validate the query payload.
    /// Validations performed:
    /// - Validate that all operations are of the same type (queries or mutations)
    /// - Validate that there is at least one operation
    /// - Other validations are delegated to the operation validator
    pub fn validate(
        self,
        document: ExecutableDocument,
    ) -> Result<ValidatedDocument, ExecutionError> {
        let operation_validator = OperationValidator::new(
            self.schema,
            self.operation_name,
            self.variables,
            document.fragments,
        );

        let operations = document
            .operations
            .iter()
            .map(|operation| operation_validator.validate_operation(operation))
            .collect::<Result<Vec<_>, _>>()?;

        let operation_typ = match &operations.first() {
            Some(operation) => {
                let same_operation_type = operations.iter().all(|op| op.typ == operation.typ);
                if same_operation_type {
                    Ok(operation.typ)
                } else {
                    Err(ExecutionError::DifferentOperationTypes)
                }
            }
            None => Err(ExecutionError::NoOperationFound),
        }?;

        Ok(ValidatedDocument {
            operations,
            operation_typ,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use async_graphql_parser::parse_query;

    #[test]
    fn argument_valid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                concert(id: 1) {
                    id
                    title
                    venue {
                        id
                        name
                    }
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn with_operation_name_valid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query ConcertById {
                concert(id: 1) {
                    id
                    title
                    venue {
                        id
                        name
                    }
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn stray_argument_invalid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                concert(id: 1, foo: "bar") { # foo is not a valid argument
                    id
                    title
                    venue {
                        id
                        name
                    }
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn unspecified_required_argument_invalid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                concert { # id argument is required here
                    id
                    title
                    venue {
                        id
                        name
                    }
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn variable_resolution_valid() {
        let schema = create_test_schema();

        let variables = create_variables(
            r#"
            {
                "concert_id": 1,
                "venue_id": 2
            }"#,
        );

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: Some(variables),
        };

        let query = r#"
            query($concert_id: Int!, $venue_id: Int!) {
                concert(id: $concert_id) { # id argument is required here
                    id
                    title
                }
                venue(id: $venue_id) {
                    id
                    name
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn variable_resolution_invalid() {
        let schema = create_test_schema();

        let variables = create_variables(r#"{ "concert_id": 2 }"#);
        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: Some(variables),
        };

        let query = r#"
            query($concert_id: Int!, $venue_id: Int!) { # venue_id is not a specified in variables
                concert(id: $concert_id) { # id argument is required here
                    id
                    title
                }
                venue(id: $venue_id) {
                    id
                    name
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn invalid_subfield() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                concert(id: 1) {
                    id
                    title
                    foobar
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    #[test]
    fn aliases_valid() {
        let schema = create_test_schema();

        let validator = DocumentValidator {
            schema: &schema,
            operation_name: None,
            variables: None,
        };

        let query = r#"
            query {
                firstConcert: concert(id: 1) {
                    id
                    headLine: title
                }
            }
        "#;

        insta::assert_debug_snapshot!(validator.validate(create_query_document(query)));
    }

    fn create_variables(variables: &str) -> Map<String, Value> {
        serde_json::from_str(variables).unwrap()
    }

    fn create_test_schema() -> Schema {
        let test_clay = r#"
            model Concert {
                id: Int = autoincrement() @pk
                title: String
                venue: Venue
            }

            model Venue {
                id: Int = autoincrement() @pk
                name: String
                concerts: Set<Concert>
            }
        "#;
        let system =
            payas_parser::build_system_from_str(test_clay, "test.clay".to_string()).unwrap();
        Schema::new(&system)
    }

    fn create_query_document(query_str: &str) -> ExecutableDocument {
        parse_query(query_str).unwrap()
    }
}
