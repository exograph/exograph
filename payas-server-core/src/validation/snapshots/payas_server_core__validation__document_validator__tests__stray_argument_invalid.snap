---
source: payas-server-core/src/validation/document_validator.rs
expression: validator.validate(create_query_document(query))
---
Err(
    StrayArguments(
        [
            "foo",
        ],
        "concert",
        Pos(3:17),
    ),
)
